msal/managed_identity.py

```diff
@@ -0,0 +1,598 @@
+                id_type=self.RESOURCE_ID, identifier=resource_id)
+        elif not client_id and not resource_id and object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.OBJECT_ID, identifier=object_id)
+        else:
+            raise ManagedIdentityError(
+                "You shall specify one of the three parameters: "
+                "client_id, resource_id, object_id")
+
+
+class _ThrottledHttpClient(ThrottledHttpClientBase):
+    def __init__(self, http_client, **kwargs):
+        super(_ThrottledHttpClient, self).__init__(http_client, **kwargs)
+        self.get = IndividualCache(  # All MIs (except Cloud Shell) use GETs
+            mapping=self._expiring_mapping,
+            key_maker=lambda func, args, kwargs: "REQ {} hash={} 429/5xx/Retry-After".format(
+                args[0],  # It is the endpoint, typically a constant per MI type
+                self._hash(
+                    # Managed Identity flavors have inconsistent parameters.
+                    # We simply choose to hash them all.
+                    str(kwargs.get("params")) + str(kwargs.get("data"))),
+                ),
+            expires_in=RetryAfterParser(5).parse,  # 5 seconds default for non-PCA
+            )(http_client.get)
+
+
+class ManagedIdentityClient(object):
+    """This API encapsulates multiple managed identity back-ends:
+    VM, App Service, Azure Automation (Runbooks), Azure Function, Service Fabric,
+    and Azure Arc.
+
+    It also provides token cache support.
+
+    .. note::
+
+        Cloud Shell support is NOT implemented in this class.
+        Since MSAL Python 1.18 in May 2022, it has been implemented in
+        :func:`PublicClientApplication.acquire_token_interactive` via calling pattern
+        ``PublicClientApplication(...).acquire_token_interactive(scopes=[...], prompt="none")``.
+        That is appropriate, because Cloud Shell yields a token with
+        delegated permissions for the end user who has signed in to the Azure Portal
+        (like what a ``PublicClientApplication`` does),
+        not a token with application permissions for an app.
+    """
+    _instance, _tenant = socket.getfqdn(), "managed_identity"  # Placeholders
+
+    def __init__(
+        self,
+        managed_identity: Union[
+            dict,
+            ManagedIdentity,  # Could use Type[ManagedIdentity] but it is deprecatred in Python 3.9+
+            SystemAssignedManagedIdentity,
+            UserAssignedManagedIdentity,
+            ],
+        *,
+        http_client,
+        token_cache=None,
+        http_cache=None,
+    ):
+        """Create a managed identity client.
+
+        :param managed_identity:
+            It accepts an instance of :class:`SystemAssignedManagedIdentity`
+            or :class:`UserAssignedManagedIdentity`.
+            They are equivalent to a dict with a certain shape,
+            which may be loaded from a json configuration file or an env var,
```

Comment:
1. It's better to capitalize `json` just like how it appears elsewhere.
2. The sentence should end with a period, not a comma.

```suggestion
            which may be loaded from a JSON configuration file or an env var.
```